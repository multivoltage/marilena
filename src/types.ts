import { MJMLParsingOptions } from "mjml-core";
export interface UserConfig {
  inputFolder?: string; // default input
  outputFolder?: string; // default output
  htmlVersion?: (emailName: string, locale: string) => string;
  textVersion?: (emailName: string, locale: string) => string;
  templateOptions?: {
    engine: SupportedEngine;
    // engineRoot will be handlebars,etc etc...
    prepareEngine: (engineRoot: unknown) => void;
  };
  locales?: string[]; // default ["en"]
  mjmlParsingOptions?: MJMLParsingOptions;
}

export type SupportedEngine = "eta" | "handlebars";

// internal use only
export type CoreConfig = UserConfig &
  Required<Pick<UserConfig, "inputFolder" | "outputFolder" | "locales">>;
