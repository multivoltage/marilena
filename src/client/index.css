html > body {
  margin: 0;
  padding: 0;
}

html,
body,
.email-frame-container,
.page,
.page-wrapper,
iframe {
  width: 100%;
  height: 100%;
}

body {
  font-family: "Roboto", sans-serif;
}

.page__email .send-to {
  position: fixed;
  top: 0;
  right: 0;
  display: flex;
  flex-direction: column;
}

.page__email .send-to input[type="email"] {
  width: 200px;
}

.send-to__modal button {
  position: absolute;
  right: 1rem;
  top: 1rem;
}

#email-frame {
  transition: width 200ms ease-in-out;
}

.email-frame-container,
.frame-settings {
  display: flex;
  justify-content: center;
}

.frame-settings {
  margin: 1rem 0;
}
