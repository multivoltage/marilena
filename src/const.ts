export const FILE_NAME_COMMON_VARIABLES = "common";
export const FILE_NAME_EMAIL_VARIABLES = "variables";
export const FILE_NAME_EMAIL_METADATA = "metadata";

export const EVENT_NAME_NEED_REFRESH_WEBSOCKET = "need_refresh";

export const CONFIG_FILE_NAME = "marilena.config.mjs";
export const SERVER_PORT = 8080;
