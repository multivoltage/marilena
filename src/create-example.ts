import path from "path";
import fs from "fs";
import createInpuFolder from "./create-example/create-input-folder";
import createOutputFolder from "./create-example/create-output-folder";
import createConfig from "./create-example/create-config";
import createMail from "./create-example/create-mail";
import createLayoutAndPartial from "./create-example/create-layout-and-partials";
import createCommonEn from "./create-example/create-common-en";
import createStyles from "./create-example/create-style";
import logger from "node-color-log";

const rootFolder = process.cwd();
const folderName = "example";
const fromPath = path.join(rootFolder, folderName);

// remove all if exist
try {
  fs.rmSync(fromPath, {
    recursive: true,
  });
} catch {}

// 1 create example folder
fs.mkdirSync(fromPath, {
  recursive: true,
});

// 2 create marilena.config.mjs
createConfig(fromPath);

// 3 create example/input example/ouuput
createInpuFolder(fromPath);
createOutputFolder(fromPath);

// 4 create example/input/welcome (en, index.html)
createMail(fromPath);

// 5 create layout and partials
createLayoutAndPartial(fromPath);

// 6 create example/input/common-en.yml
createCommonEn(fromPath);

// 7 create example/input/styles.css
createStyles(fromPath);

logger.debug(
  'I created a basic working example for you :). Check folder "example"',
);
