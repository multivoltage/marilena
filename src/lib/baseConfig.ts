import { Config } from "../types";

const config: Config = {
  inputFolder: "input",
  outputFolder: "output",
  locales: ["en"],
};

export default config;
