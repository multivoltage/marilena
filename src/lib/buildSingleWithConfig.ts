import fs from "fs";
import path from "path";
import { Config } from "../types";
import { inputOutputHtml } from "./inputOutputHtml";
import loadVariables from "./loadVariables";
import logger from "node-color-log";

export async function buildSingle(config: Config, emailName: string) {
  const {
    inputFolder,
    outputFolder,
    locales,
    mjmlParsingOptions,
    textVersion,
  } = config;

  const inputFolderPath = path.join(inputFolder);
  const outputFolderPath = path.join(outputFolder);

  const inputEmailFilePath = path.join(
    inputFolderPath,
    emailName,
    `index.html`
  );
  const mjmlTemplate = fs.readFileSync(inputEmailFilePath, "utf-8");

  // create folder for each email and locale if not exist
  for (const locale of locales) {
    const emailFolderPath = path.join(outputFolderPath, emailName, locale);
    if (!fs.existsSync(emailFolderPath)) {
      fs.mkdirSync(emailFolderPath, { recursive: true });
    }

    const variables = loadVariables({ config, locale, emailName });
    const html = await inputOutputHtml({
      inputHtml: mjmlTemplate,
      variables,
      templateOptions: config.templateOptions,
      mjmlParsingOptions,
      isTextVersion: false,
    });

    const folderEmailPathLang = path.join(outputFolderPath, emailName, locale);
    if (!fs.existsSync(folderEmailPathLang)) {
      fs.mkdirSync(folderEmailPathLang, { recursive: true });
    }

    try {
      fs.writeFileSync(path.join(folderEmailPathLang, "index.html"), html, {
        encoding: "utf-8",
      });
      logger.info("writed", emailName, locale);
    } catch (e) {
      logger.error("failed writing", emailName, e);
    }

    // text version
    if (textVersion) {
      try {
        const htmlOnlyText = await inputOutputHtml({
          inputHtml: mjmlTemplate,
          variables,
          templateOptions: config.templateOptions,
          mjmlParsingOptions,
          isTextVersion: true,
        });
        fs.writeFileSync(
          path.join(folderEmailPathLang, textVersion(emailName, locale)),
          htmlOnlyText,
          {
            encoding: "utf-8",
          }
        );
        logger.info("writed", emailName, locale, "text version");
      } catch (e) {
        logger.error("failed writing text version", emailName, e);
      }
    }
  }
}
