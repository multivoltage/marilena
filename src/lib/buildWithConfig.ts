import fs from "fs";
import path from "path";
import { CoreConfig } from "../types";
import { buildSingle } from "./buildSingleWithConfig";
import { getPathConfig, isEmailDirectory } from "../utils";

export async function build(config: CoreConfig) {
  const { inputFolder, outputFolder } = config;

  const inputFolderPath = path.resolve(getPathConfig(), "..", inputFolder);
  const outputFolderPath = path.resolve(getPathConfig(), "..", outputFolder);

  // create folder if not exist
  if (!fs.existsSync(outputFolderPath)) {
    fs.mkdirSync(outputFolderPath, { recursive: true });
  }

  let list: string[];

  try {
    const files = fs.readdirSync(inputFolderPath, { withFileTypes: true });
    list = files
      .filter((f) => isEmailDirectory(inputFolderPath, f))
      .map((f) => f.name);
  } catch (e) {
    console.error(e);
    list = [];
  }

  for (const emailName of list) {
    await buildSingle(config, emailName);
  }
}
