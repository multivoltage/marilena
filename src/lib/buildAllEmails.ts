import { loadConfig } from "../utils.js";
import { build } from "./buildWithConfig.js";

async function buildAllEmail() {
  const config = await loadConfig();
  build(config);
}

buildAllEmail();
