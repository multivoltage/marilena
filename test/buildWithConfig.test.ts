import { beforeAll, describe, expect, test } from "@jest/globals";
import { CoreConfig } from "../src/types";
import { build } from "../src/lib/buildWithConfig";
import fs from "node:fs";
import path from "node:path";
import { rimrafSync } from "rimraf";
import { Eta } from "eta";

const testEmailName = "hello"; // we have only hello.html inside inputs

describe("writer correctly email files", () => {
  beforeAll(() => {
    // delete all output test folder
    rimrafSync("test/*/output", { glob: true });
  });

  test("create email with: [engine: NO, suffix: html, locale: en]", async () => {
    const config: CoreConfig = {
      inputFolder: "test/basic_1/input",
      outputFolder: "test/basic_1/output",
      locales: ["en"],
    };

    await build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "index.html"),
      { encoding: "utf-8" },
    );

    expect(outpoutEn).toMatchSnapshot();
  });

  test("create email with: [engine: NO, suffix: html, locale: en] - custom htmlVersion", async () => {
    const config: CoreConfig = {
      inputFolder: "test/basic_1/input",
      outputFolder: "test/basic_1/output",
      htmlVersion: (emailName, locale) => `${emailName}-${locale}.html`,
      locales: ["en"],
    };

    await build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "hello-en.html"),
      { encoding: "utf-8" },
    );

    expect(outpoutEn).toMatchSnapshot();
  });

  test("create email with: [engine: eta, suffix: html, locale: en]", async () => {
    const config: CoreConfig = {
      inputFolder: "test/eta_1/input",
      outputFolder: "test/eta_1/output",
      locales: ["en"],
      templateOptions: {
        engine: "eta",
        prepareEngine: (engine: any) => {
          const eta = engine as Eta;
          eta.configure({
            views: path.resolve(__dirname, "eta_1/input/partials"),
          });

          eta.loadTemplate("@partial_1", "this is programmatly partial");
        },
      },
    };

    await build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "index.html"),
      { encoding: "utf-8" },
    );

    fs.rmSync(config.outputFolder, { recursive: true });

    expect(outpoutEn).toMatchSnapshot();
  });

  test("create email with: [engine: handlebars, suffix: html, locale: en]", async () => {
    const config: CoreConfig = {
      inputFolder: "test/handlebars_1/input",
      outputFolder: "test/handlebars_1/output",
      locales: ["en"],
      templateOptions: {
        engine: "handlebars",
        prepareEngine: (h: any) => {
          h.registerPartial("myPartial", "partial with {{ user }}");
        },
      },
    };

    await build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "index.html"),
      { encoding: "utf-8" },
    );

    fs.rmSync(config.outputFolder, { recursive: true });

    expect(outpoutEn).toMatchSnapshot();
  });

  test("create text version email", async () => {
    const config: CoreConfig = {
      inputFolder: "test/text_version/input",
      outputFolder: "test/text_version/output",
      locales: ["en"],
      textVersion: (emailName, locale) => `${emailName}-${locale}.txt`,
    };

    await build(config);

    const outpoutTextVersion = fs.readFileSync(
      path.join(
        config.outputFolder,
        testEmailName,
        "en",
        `${testEmailName}-en.txt`,
      ),
      { encoding: "utf-8" },
    );

    fs.rmSync(config.outputFolder, { recursive: true });

    expect(outpoutTextVersion).toMatchSnapshot();
  });

  test("create email with unsupported engine should throw error", async () => {
    expect.assertions(1);

    const config: CoreConfig = {
      inputFolder: "test/handlebars_1/input",
      outputFolder: "test/handlebars_1/output",
      locales: ["en"],
      templateOptions: {
        engine: "fake-engine" as any, // we use any since thid file is in ts and a consumer can use a normal mjs file without import types
        prepareEngine: () => {},
      },
    };

    try {
      await build(config);
    } catch (e) {
      expect(e.message).toContain(
        "engine fake-engine not supported. Options available are",
      );
    }
  });
});
