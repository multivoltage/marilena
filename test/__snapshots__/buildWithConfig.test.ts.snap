// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writer correctly email files create email with: [engine: NO, suffix: html, locale: en] 1`] = `
"<mjml>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-text>test</mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
"
`;

exports[`writer correctly email files create email with: [engine: eta, suffix: html, locale: en] 1`] = `
"<mjml>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-text>hello en</mj-text>
        <mj-text>this is programmatly partial</mj-text>
        <mj-text>age 32</mj-text>
        <mj-text>this is file partial</mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
"
`;

exports[`writer correctly email files create email with: [engine: handlebars, suffix: html, locale: en] 1`] = `
"<mjml>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-text>hello diego tonini has age 12</mj-text>
        <mj-text>partial with diego tonini</mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
"
`;

exports[`writer correctly email files create text version email 1`] = `"only text version"`;
